### PR TITLE
Fix qml.grad so that the returned gradient always matches the cost function return type if only a single argument is differentiated

### PR DESCRIPTION
**Pull request by josh146**
**Context:**

I realised yesterday that `autograd.grad` behaves differently depending if the `argnum` argument is a sequence or an integer:

```python
>>> import autograd
>>> from autograd import numpy as np
>>> autograd.grad(np.sin, argnum=0)(0.6)
0.8253356149096783
>>> autograd.grad(np.sin, argnum=[0])(0.6)
(array(0.82533561),)
```

We never take this into account in PennyLane, which can lead to gradients often being returned as tuple of size 1.

**Description of the Change:**

* If the inferred `argnum` argument is a sequence of length 1, only the contained integer is passed to `autograd.grad`.

**Benefits:**

* The return of `qml.grad` should now _always_ match the return type of the function being differentiated.

* A lot of superfluous `[0]` could be removed from the tests (e.g., `qml.grad(circuit)(x)[0]`)

**Possible Drawbacks:**

* The optimizers have to add in the redundant tuple where needed so that the gradient remains iterable. But this is a easy change, and leads to an intuitive UI, so maybe better?

* Some of the demos might need to be updated to remove unneeded `[0]` after gradient computations. But I think very few demos actually show the gradients.

**Related GitHub Issues:** n/a

### CODE REVIEW

File: tests/tape/interfaces/test_qnode_autograd.py [hunk 1]
```diff
@@ -594,6 +592,127 @@ def circuit():
...
+
+        def cost(a, b, c, weights):
+            return circuit2(circuit1(a, b, c), weights)
+
+        grad_fn = qml.grad(cost)
+
+        # Set the first parameter of circuit1 as non-differentiable.
+        a = np.array(0.4, requires_grad=False)
+
+        # the remaining free parameters are all differentiable
```
Comment (Thenerdstation) [inline tests/tape/interfaces/test_qnode_autograd.py:676]:
```suggestion
        # The remaining free parameters are all differentiable.
```

File: tests/tape/interfaces/test_qnode_autograd.py [hunk 2]
```diff
@@ -594,6 +592,127 @@ def circuit():
...
+        # np.hstack 'flattens' the ragged gradient array allowing it
+        # to be compared with the expected result
+        assert np.allclose(np.hstack(res), expected, atol=tol, rtol=0)
+
+        if diff_method != "backprop":
+            # Check that the gradient was computed
+            # for all parameters in circuit2
+            assert circuit2.qtape.trainable_params == {0, 1, 2, 3}
+
+            # check that the parameter-shift rule was not applied
```
Comment (Thenerdstation) [inline tests/tape/interfaces/test_qnode_autograd.py:712]:
```suggestion
            # Check that the parameter-shift rule was not applied
```

File: tests/tape/interfaces/test_qnode_autograd.py [hunk 3]
```diff
@@ -594,6 +592,127 @@ def circuit():
...
+            # to the first parameter of circuit1
```
Comment (Thenerdstation) [inline tests/tape/interfaces/test_qnode_autograd.py:713]:
```suggestion
            # to the first parameter of circuit1.
```

File: pennylane/optimize/gradient_descent.py [hunk 4]
```diff
@@ -127,6 +127,9 @@ def compute_grad(objective_fn, args, kwargs, grad_fn=None):
         grad = g(*args, **kwargs)
         forward = getattr(g, "forward", None)
 
+        if len(args) == 1:
+            grad = (grad,)
+
```
Comment (Thenerdstation) [inline pennylane/optimize/gradient_descent.py:132]: We have to turn grad into a tuple because we're reversing what we did to argnum earlier, right?

**Review by Thenerdstation — APPROVED** [review]

File: pennylane/optimize/gradient_descent.py (hunk 4 above)
Comment (josh146) [reply pennylane/optimize/gradient_descent.py:132]: Yep, the optimizer logic currently assumes `grad` is always iterable with respect to argument number.

I figured, easier to leave this logic intact and simply re-tuple the grad

Comment (Thenerdstation) [reply pennylane/optimize/gradient_descent.py:132]: Agreed.
